gr-zeromq/lib/base_impl.cc

```diff
@@ -43,6 +47,42 @@ base_impl::base_impl(int type,
     if (major < 3) {
         d_timeout *= 1000;
     }
+
+    /* Set ZMQ_LINGER so socket won't infinitely block during teardown */
+#if USE_NEW_CPPZMQ_SET_GET
+    d_socket.set(zmq::sockopt::linger, LINGER_DEFAULT);
+#else
+    d_socket.setsockopt(ZMQ_LINGER, &LINGER_DEFAULT, sizeof(LINGER_DEFAULT));
+#endif
+
+    /* Set high watermark */
+    if (hwm >= 0) {
+#if USE_NEW_CPPZMQ_SET_GET
+        if (sink) {
+            d_socket.set(zmq::sockopt::sndhwm, hwm);
+        } else {
+            d_socket.set(zmq::sockopt::rcvhwm, hwm);
+        }
+#else
+#ifdef ZMQ_SNDHWM
+        if (sink) {
+            d_socket.setsockopt(ZMQ_SNDHWM, &hwm, sizeof(hwm));
+        } else {
+            d_socket.setsockopt(ZMQ_RCVHWM, &hwm, sizeof(hwm));
```

Comment:
ZMQ_SNDHWM and ZMQ_RCVHWM were introduced at the same time, replacing ZMQ_HWM. Maybe checking version numbers would work, too. I think this is good enough for now since even Ubuntu 18.04 has a newer version.